sources/web/datalab/polymer/components/item-list/item-list.ts

```diff
@@ -90,93 +96,138 @@ class ItemListElement extends Polymer.Element {
-      for (let i = 0; i < this.rows.length; ++i) {
-        this.set('rows.' + i + '.selected', false);
-      }
-      this.set('rows.' + index + '.selected', true);
 
-      // This is now the only selected element.
-      this._selectedElements = [rowElement];
-    } else {
+    // If shift key is pressed and we had saved the last selected index, select
+    // all items from this index till the last selected.
+    if (e.shiftKey && this._lastSelectedIndex !== -1 && this.selectedIndices.length > 0) {
+      this._unselectAll();
+      const start = Math.min(this._lastSelectedIndex, index);
+      const end = Math.max(this._lastSelectedIndex, index);
+      for (let i = start; i <= end; ++i) {
+        this._selectItem(i);
+      }
+    }
+    
+    // If ctrl key is pressed, toggle its selection.
+    else if (e.ctrlKey) {
       if (this.rows[index].selected === false) {
-        // Remove this element from the selected elements list if it's being unselected
-        const i = this._selectedElements.indexOf(rowElement);
-        if (i > -1) {
-          this.splice('_selectedElements', i, 1);
+        this._selectItem(index);
+      } else {
+        this._unselectItem(index);
+      }
+    }
+    
+    // No modifier keys are pressed, proceed normally to select/unselect the item.
+    else {
+      // If the clicked element is the checkbox, the checkbox already toggles selection in
+      // the UI, so change the item's selection state to match the checkbox's new value.
+      // Otherwise, select this element, unselect all others.
+      if (target.tagName === 'PAPER-CHECKBOX') {
+        if (this.rows[index].selected === false) {
+          // Remove this element from the selected elements list if it's being unselected
+          this._unselectItem(index);
+        } else {
+          // Add this element to the selected elements list if it's being selected,
+          this._selectItem(index);
         }
       } else {
-        // Add this element to the selected elements list if it's being selected,
-        this.push('_selectedElements', rowElement);
+        this._unselectAll();
+        this._selectItem(index);
       }
     }
-    const ev = new ItemClickEvent('itemSelectionChanged', { detail: {index: index} });
-    this.dispatchEvent(ev);
-  }
 
-  /**
-   * Given an element inside a row in the list, finds the parent row element.
-   */
-  _getRowElementFromChild(childElement: HTMLElement): HTMLElement {
-    let currentElement = childElement;
-    while (currentElement.tagName !== 'PAPER-ITEM' && !currentElement.classList.contains('row'))
-      if (currentElement.parentElement)
-        currentElement = currentElement.parentElement;
-      else
-        // This should not happen
-        throw new Error('Could not find parent row element for: ' + childElement.tagName);
-    return currentElement;
+    // Save this index to enable multi-selection using shift later.
+    this._lastSelectedIndex = index;
```

Comment:
In all of my examples, the first click was a single-select, the second click was an add-to select (control or command), and the third click was a range-select (shift). So after the first two clicks, items 2 and 4 are selected, and the result after the next click depends on which was selected first and what platform it is on. Finder appears to remember the last item clicked and uses that for the range, as your code does, but Nautilus appears to remember the first item clicked that is still selected - or something like that, I did not try to fully characterize.